packages/edit-site/src/components/post-list/index.js

```diff
@@ -39,100 +40,123 @@ const { useLocation, useHistory } = unlock( routerPrivateApis );
 			}
+
 			const { getEditedEntityRecord } = select( coreStore );
-			const dataviewRecord = getEditedEntityRecord(
+			return getEditedEntityRecord(
 				'postType',
 				'wp_dataviews',
 				Number( activeView )
 			);
-			return dataviewRecord;
 		},
 		[ activeView, isCustom ]
 	);
-	const { editEntityRecord } = useDispatch( coreStore );
-
-	const customView = useMemo( () => {
-		const storedView =
-			editedViewRecord?.content &&
-			JSON.parse( editedViewRecord?.content );
-		if ( ! storedView ) {
-			return storedView;
-		}
-
-		return {
-			...storedView,
-			layout: defaultLayouts[ storedView?.type ]?.layout,
-		};
-	}, [ editedViewRecord?.content ] );
-
-	const setCustomView = useCallback(
-		( viewToSet ) => {
-			editEntityRecord(
-				'postType',
-				'wp_dataviews',
-				editedViewRecord?.id,
-				{
-					content: JSON.stringify( viewToSet ),
+	const [ view, setView ] = useState( () => {
+		if ( isCustom === 'true' ) {
+			return (
+				getCustomView( editedEntityRecord ) ?? {
+					type: layout ?? LAYOUT_TABLE,
 				}
 			);
-		},
-		[ editEntityRecord, editedViewRecord?.id ]
-	);
+		}
+		return (
+			getDefaultView( defaultViews, activeView ) ?? {
+				type: layout ?? LAYOUT_TABLE,
+			}
+		);
+	} );
 
-	const setDefaultViewAndUpdateUrl = useCallback(
-		( viewToSet ) => {
-			if ( viewToSet.type !== view?.type ) {
-				const { params } = history.getLocationWithParams();
+	const setViewWithUrlUpdate = useCallback(
+		( newView ) => {
+			const { params } = history.getLocationWithParams();
+
+			if ( newView.type !== params?.layout ) {
 				history.push( {
 					...params,
-					layout: viewToSet.type,
+					layout: newView.type,
 				} );
 			}
-			setView( viewToSet );
+
+			setView( newView );
+
+			if ( isCustom === 'true' && editedEntityRecord?.id ) {
+				editEntityRecord(
+					'postType',
+					'wp_dataviews',
+					editedEntityRecord?.id,
+					{
+						content: JSON.stringify( newView ),
+					}
+				);
+			}
 		},
-		[ history, view?.type ]
+		[ history, isCustom, editedEntityRecord?.id ]
 	);
 
-	if ( isCustom === 'false' ) {
-		return [ view, setDefaultViewAndUpdateUrl ];
-	} else if ( isCustom === 'true' && customView ) {
-		return [ customView, setCustomView ];
-	}
-	// Loading state where no the view was not found on custom views or default views.
-	return [ DEFAULT_VIEWS[ postType ][ 0 ].view, setDefaultViewAndUpdateUrl ];
+	// When layout URL param changes, update the view type
+	// without affecting any other config.
+	useEffect( () => {
+		setView( ( prevView ) => ( { ...prevView, type: layout } ) );
+	}, [ layout ] );
+
+	// When activeView or isCustom URL parameters change,
+	// reset the view & update the layout URL param to match the view's type.
+	useEffect( () => {
+		let newView;
+		if ( isCustom === 'true' ) {
+			newView = getCustomView( editedEntityRecord );
+		} else {
+			newView = getDefaultView( defaultViews, activeView );
+		}
+
+		if ( newView ) {
+			setViewWithUrlUpdate( newView );
```

Comment:
I think when we switch to a custom view, as here we call setViewWithUrlUpdate which has this code `if ( isCustom === 'true' && editedEntityRecord?.id ) { editEntityRecord(`, we will endup calling get editEntityRecord just because of loading the view which may trigger a rerender. Not a big issue.

Reply:
That's right. I've looked at this but don't see a way to improve it at the moment. Custom views are an experimental feature and this isn't a big issue, so I'll go ahead with this PR anyway because it provides a lot of value.

I'd be happy to pair/look at this in a follow-up.